drivers/hitachi/projector/cp_tw_series_basic.cr

```diff
@@ -0,0 +1,230 @@
+
+  @recover_power : PlaceOS::Driver::Proxy::Scheduler::TaskWrapper? = nil
+  @recover_input : PlaceOS::Driver::Proxy::Scheduler::TaskWrapper? = nil
+  # nil by default (allows manual on and off)
+  @power_target : Bool? = nil
+  @input_target : String? = nil
+
+  def on_load
+    # Response time is slow
+    # and as a make break device it may take time
+    # to actually setup the connection with the projector
+    queue.delay = 100.milliseconds
+    queue.timeout = 5.seconds
+    queue.retries = 3
+
+    # Meta data for inquiring interfaces
+    self[:type] = :projector
+  end
+
+  def connected
+    schedule.every(50.seconds, true) { poll_1 }
+    schedule.every(10.minutes, true) { poll_2 }
+  end
+
+  def poll_1
+    power?(priority: 0).get
+    if self[:power]?.try &.as_bool
+      input?(priority: 0)
+      audio_mute?(priority: 0)
+      picture_mute?(priority: 0)
+      freeze?(priority: 0)
+    end
+  end
+
+  def poll_2
+    lamp?(priority: 0)
+    filter?(priority: 0)
+    error?(priority: 0)
+  end
+
+  def disconnected
+    schedule.clear
+    @recover_power = nil
+    @recover_input = nil
+  end
+
+  def power(state : Bool)
+    @power_target = state
+    if state
+      logger.debug { "-- requested to power on" }
+      do_send("BA D2 01 00 00 60 01 00", name: "power")
+    else
+      logger.debug { "-- requested to power off" }
+      do_send("2A D3 01 00 00 60 00 00", name: "power")
+    end
+    power?
+  end
+
+  INPUTS = {
+    "hdmi"  => "0E D2 01 00 00 20 03 00",
+    "hdmi2" => "6E D6 01 00 00 20 0D 00",
+  }
+
+  def switch_to(input : String)
+    @input_target = input
+    do_send(INPUTS[input], name: "input")
+    input?
+  end
+
+  def mute(
+    state : Bool = true,
+    index : Int32 | String = 0,
+    layer : MuteLayer = MuteLayer::AudioVideo
+  )
+    mute_video(state) if layer.video? || layer.audio_video?
+    mute_audio(state) if layer.audio? || layer.audio_video?
+  end
+
+  def mute_video(state : Bool = true)
+    if state
+      do_send("6E F1 01 00 A0 20 01 00", name: "mute_video")
+    else
+      do_send("FE F0 01 00 A0 20 00 00", name: "mute_video")
+    end
+    picture_mute?
+  end
+
+  def mute_audio(state : Bool = true)
+    if state
+      do_send("D6 D2 01 00 02 20 01 00", name: "mute_audio")
+    else
+      do_send("46 D3 01 00 02 20 00 00", name: "mute_audio")
+    end
+    audio_mute?
+  end
+
+  QueryRequests = {
+    power:        "19 D3 02 00 00 60 00 00",
+    input:        "CD D2 02 00 00 20 00 00",
+    error:        "D9 D8 02 00 20 60 00 00",
+    freeze:       "B0 D2 02 00 02 30 00 00",
+    audio_mute:   "75 D3 02 00 02 20 00 00",
+    picture_mute: "CD F0 02 00 A0 20 00 00",
+    lamp:         "C2 FF 02 00 90 10 00 00",
+    filter:       "C2 F0 02 00 A0 10 00 00",
+  }
+  {% for name, data in QueryRequests %}
+    @[Security(Level::Administrator)]
+    def {{name.id}}?(**options)
+      do_send({{data}}, **options, name: {{name.id.stringify}} + '?')
+    end
+  {% end %}
```

Comment:
It might make it a bit clearer and less prone to error to have all of the messages collected in a single place.
the do send method [in this driver](https://github.com/PlaceOS/drivers/blob/8c505742bbac53180d3ecdbff842e160277fbebf/drivers/panasonic/projector/nt_control.cr#L247) is pretty neat, and given that it's a  NamedTuple, invalid keys to the hash are checked at compile time.


Then these lines could look something like this

```suggestion
  QueryRequests = %i(power input error freeze audio_mute picture_mute lamp filter)
  {% for name in QueryRequests %}
    @[Security(Level::Administrator)]
    def {{name.id}}?(**options)
      do_send({{ name }}, **options, name: {{name.id.stringify}} + '?')
    end
  {% end %}
```

Reply:
kim mentioned the enum manner of doing things as well, which is good too

Reply:
addressed with https://github.com/PlaceOS/drivers/pull/73/commits/07bb5c0232a9c1012de0b76be39a07ff8c26b803